### PR TITLE
grandpa-rpc: use FinalityProofProvider to check finality for rpc

### PR DESCRIPTION
**Pull request by octol**
#5732  introduced a subscription based API for streaming finality proofs (justifications) as they become available.

This PR adds an on-demand API that clients can call to catch up on the stream of proofs if they need to fill in missing proofs. For example after a period of being offline.

The implementation is based on shared the existing `FinalityProofProvider` with the Grandpa RPC handler which is then used to query for finality proofs.

TODO
- [x] Figure out how to handle the authorities set ID that the API currently expects to be passed in.
- [x] Refactor away the unnecessary encode/decode step that the initial implementation does.
- [x] Tests.

Part of #6066 

polkadot companion: https://github.com/paritytech/polkadot/pull/1654

### CODE REVIEW

File: client/finality-grandpa/rpc/src/lib.rs [hunk 1]
```diff
@@ -80,39 +84,51 @@ pub trait GrandpaApi<Notification> {
 		metadata: Option<Self::Metadata>,
 		id: SubscriptionId
 	) -> jsonrpc_core::Result<bool>;
+
+	/// Prove finality for the hash, given a last known finalized hash.
+	#[rpc(name = "grandpa_proveFinality")]
+	fn prove_finality(
+		&self,
+		last_finalized: Hash,
```
[the remark this answers is not in the document]
Comment (octol) [reply client/finality-grandpa/rpc/src/lib.rs:89]: Ah ok, so just let the caller provide the `Hash` instead of as now use `self.backend.blockchain().info().best_hash`?

File: client/finality-grandpa/src/finality_proof.rs [hunk 2]
```diff
@@ -180,7 +180,39 @@ impl<B, Block: BlockT> FinalityProofProvider<B, Block>
...
+		Block: BlockT,
+		NumberFor<Block>: BlockNumberOps,
+		B: Backend<Block> + Send + Sync + 'static,
+{
+	fn prove_finality(
+		&self,
+		last_finalized: Block::Hash,
+		authorities_set_id: u64,
+	) -> Result<Option<Vec<u8>>, ClientError> {
+		use sp_blockchain::HeaderBackend;
```
[the remark this answers is not in the document]
Comment (octol) [reply client/finality-grandpa/src/finality_proof.rs:198]: Good point. Do we want to mirror the output in the pubsub API. Send a `GrandpaJustification` for every block in the range? In `prove_finality` we loop over `blockchain.justification(current_id)`, can we send that?

[the remark this answers is not in the document]
Comment (andresilva) [reply client/finality-grandpa/src/finality_proof.rs:198]: Not sure what you mean by this, my comment is a further explanation of what I said above that this needs to take two hashes as input, i.e. the user of the API says: I want you to prove that block #30 is final and my starting point is #10. The finality proof will contain all the intermediate `FinalityProofFragment`s that are required to prove everything in the given range.

File: bin/node/cli/src/service.rs [hunk 3]
```diff
@@ -102,14 +105,17 @@ pub fn new_partial(config: &Configuration) -> Result<sc_service::PartialComponen
...
 
+
 	let (rpc_extensions_builder, rpc_setup) = {
 		let (_, grandpa_link, babe_link) = &import_setup;
 
 		let justification_stream = grandpa_link.justification_stream();
 		let shared_authority_set = grandpa_link.shared_authority_set().clone();
 		let shared_voter_state = grandpa::SharedVoterState::empty();
+		let finality_proof_provider =
+			GrandpaFinalityProofProvider::new_for_service(backend.clone(), client.clone());
```
Comment (octol) [inline bin/node/cli/src/service.rs:116]: My understanding is that it should be fine to move the creation of the finality proof provider from `new_full_base` to `new_partial` ... ?

File: bin/node/cli/src/service.rs [hunk 4]
```diff
@@ -102,14 +105,17 @@ pub fn new_partial(config: &Configuration) -> Result<sc_service::PartialComponen
 
 	let import_setup = (block_import, grandpa_link, babe_link);
 
+
```
Comment (andresilva) [inline bin/node/cli/src/service.rs:108]:
```suggestion
```

File: client/finality-grandpa/src/lib.rs [hunk 5]
```diff
@@ -125,7 +125,7 @@ mod until_imported;
 mod voting_rule;
 
 pub use authorities::SharedAuthoritySet;
-pub use finality_proof::{FinalityProofProvider, StorageAndProofProvider};
+pub use finality_proof::{FinalityProofProvider, StorageAndProofProvider, FinalityProofFragment};
```
Comment (andresilva) [inline client/finality-grandpa/src/lib.rs:128]:
```suggestion
pub use finality_proof::{FinalityProofFragment, FinalityProofProvider, StorageAndProofProvider};
```

File: bin/node/rpc/src/lib.rs [hunk 6]
```diff
@@ -115,6 +120,8 @@ pub fn create_full<C, P, SC>(
 	C::Api: BlockBuilder<Block>,
 	P: TransactionPool + 'static,
 	SC: SelectChain<Block> +'static,
+	B: Send + Sync + 'static + sc_client_api::Backend<Block>,
+	<B as sc_client_api::Backend<Block>>::State: sp_state_machine::Backend<BlakeTwo256>,
```
Comment (andresilva) [inline bin/node/rpc/src/lib.rs:124]: Why is this needed?

**Review by andresilva — CHANGES_REQUESTED** [review]
overall lgtm, just a nit regarding some bounds that don't seem necessary.

File: bin/node/rpc/src/lib.rs (hunk 6 above)
Comment (andresilva) [inline bin/node/rpc/src/lib.rs:124]:
```suggestion
	B: sc_client_api::Backend<Block> + Send + Sync + 'static,
	B::State: sp_state_machine::Backend<sp_runtime::traits::HashFor<Block>>,
```

Comment (andresilva) [reply bin/node/rpc/src/lib.rs:124]: I think the reason we need to specify the state backend here is because `Block` isn't a generic type parameter and is actually the concrete block type. Either way doing it this way is more generic since we don't tie the bound to the hashing we currently use.

File: bin/node/rpc/src/lib.rs [hunk 7]
```diff
@@ -47,6 +49,7 @@ use sp_blockchain::{Error as BlockChainError, HeaderMetadata, HeaderBackend};
 use sp_consensus::SelectChain;
 use sp_consensus_babe::BabeApi;
 use sp_transaction_pool::TransactionPool;
+use sp_runtime::traits::BlakeTwo256;
```
Comment (andresilva) [inline bin/node/rpc/src/lib.rs:52]: This can be removed if we do the change below I suggested.

File: bin/node/rpc/Cargo.toml [hunk 8]
```diff
@@ -30,5 +30,7 @@ sp-block-builder = { version = "2.0.0-rc6", path = "../../../primitives/block-bu
 sp-blockchain = { version = "2.0.0-rc6", path = "../../../primitives/blockchain" }
 sp-consensus = { version = "0.8.0-rc6", path = "../../../primitives/consensus/common" }
 sp-consensus-babe = { version = "0.8.0-rc6", path = "../../../primitives/consensus/babe" }
+sp-runtime = { version = "2.0.0-rc6", path = "../../../primitives/runtime" }
+sp-state-machine = { version = "0.8.0-rc6", path = "../../../primitives/state-machine" }
```
Comment (andresilva) [inline bin/node/rpc/Cargo.toml:34]: I think this one isn't needed anymore.

File: client/finality-grandpa/src/finality_proof.rs [hunk 9]
```diff
@@ -180,7 +180,30 @@ impl<B, Block: BlockT> FinalityProofProvider<B, Block>
...
+}
 
+impl<B, Block> FinalityProofProvider<B, Block>
+	where
+		Block: BlockT,
+		NumberFor<Block>: BlockNumberOps,
+		B: Backend<Block> + Send + Sync + 'static,
+{
+	/// Return finality proofs for the given authorities set id, if it is provided, otherwise the
+	/// current one will be used.
```
Comment (andresilva) [inline client/finality-grandpa/src/finality_proof.rs:192]:
```suggestion
	/// Prove finality for the range (begin; end] hash. Returns None if there are no finalized blocks
	/// unknown in the range.
```

**Review by andresilva — APPROVED** [review]
lgtm. just some last nits.

**Comment by andresilva** [comment]
I think this is ready to merge but doing the last review and looking at the API we expose got me thinking that it is rather cumbersome to use (we are basically just exposing the existing `FinalyProofProvider` API). Our current API is:

```
fn prove_finality(
  begin: Hash, 
  end: Hash, 
  authorities_set_id: u64,
) -> Option<Vec<FinalityProof>>>;
```

It requires that a user of the API keep track of the authorities set id and also of block ranges spanning a given set. I think we should be able to implement the following API:

```
fn prove_finality(block: Hash) -> Option<FinalityProof>;
```

The way this API would work is that it would always generate a finality proof based on the justification from the last block in the set (which is always guaranteed to be persisted).

If a user wants to prove finality for block #52 we figure out what the set id is for that block and what the last block in the session was. Then we have all the data needed for `FinalityProofProvider` to generate the proof. I think the only thing we need to do is to start persisting the block number of the last block of the authority set so we get a `Vec<BlockNumber>`. Given any block number we could then perform binary search on this array to figure out what the last block in the set was (and implicitly the set id given by the index in the array).

For some service that is trying to sync and check for finality they would:
- `prove_finality(#1)`, for which they would get a finality proof that proves finality of all blocks up until the last block in the set;
- this proof will include all of the block headers, so they can just import them if needed (although it might make sense to make the inclusion in the proof optional and just provide block hashes);
- they will also now be able to tell what the first block in the next authority set is (i.e. last block in previous finality proof + 1);
- they loop this process until they get `None` finality proof, which means they'll be on the latest authority set;
- afterwards they can start listening to the live justifications using the pubsub API.

**Comment by andresilva** [comment]
bot merge
